packages/edit-site/src/store/reducer.js

```diff
@@ -72,25 +72,32 @@ export function settings( state = {}, action ) {
  * Reducer keeping track of the currently edited Post Type,
  * Post Id and the context provided to fill the content of the block editor.
  *
- * @param {Object} state  Current state.
+ * @param {Array}  state  Current state history.
  * @param {Object} action Dispatched action.
  *
- * @return {Object} Updated state.
+ * @return {Array} Updated state.
  */
-export function editedPost( state = {}, action ) {
+export function editedPost( state = [], action ) {
 	switch ( action.type ) {
 		case 'SET_TEMPLATE':
 		case 'SET_PAGE':
-			return {
-				type: 'wp_template',
-				id: action.templateId,
-				page: action.page,
-			};
+			return [
+				{
+					type: 'wp_template',
+					id: action.templateId,
+					page: action.page,
+				},
+			];
 		case 'SET_TEMPLATE_PART':
-			return {
-				type: 'wp_template_part',
-				id: action.templatePartId,
-			};
+			return [
+				...state,
```

Comment:
This is a neat solution! We'll have to have both `setTemplatePart` which overwrites history and e.g. `pushTemplatePart` in order to implement @jameskoster's feedback, though.